### PR TITLE
Optimize custom all reduce

### PR DESCRIPTION
**Pull request by iotamudelta**
* remove volatile keywords (they have a different meaning on CUDA than HIP)
* optimize and scope locking
* increase cutoff for custom all reduce to 16 MB (based on perf data)
* tune block number for MI300X
* roughly 10% performance increase for non-latency bound sizes, cutoff vis-a-vis RCCL at 16 MB
* while there, add missing MPI_Finalize() to test

### CODE REVIEW

**Comment by gshtras** [comment]
Looks good. Just need to fix the linters

**Comment by dllehr-amd** [comment]
Hey Folks..wanted to make a quick comment. This CAR will require rocm 6.2 to compile.  the scoping intrinsics are introduced in LLVM in the 6.2 release. So we may see compile errors on `__MEMORY_SCOPE_DEVICE` etc.

File: csrc/custom_all_reduce.cuh [hunk 1]
```diff
@@ -165,36 +174,38 @@ DINLINE void start_sync(const RankSignals& sg, volatile Signal* self_sg,
...
   __syncthreads();
   // eliminate the case that prior writes are not visible after signals become
   // visible. Note that I did not managed to make this happen through a lot of
   // testing. Might be the case that hardware provides stronger guarantee than
   // the memory model.
-  uint32_t flag = self_sg->_flag[blockIdx.x] + 1;
   if (threadIdx.x < ngpus) {
+    // reset flag for next time
+    __scoped_atomic_store_n(&self_sg->start[blockIdx.x][threadIdx.x], 0,
+                            __ATOMIC_RELAXED, __MEMORY_SCOPE_DEVICE);
```
Comment (wenkaidu) [inline csrc/custom_all_reduce.cuh:194]: Original implementation using resetting flag which is prone to race condition. Thus we have seen occasional hang during long running workload. self_sg->_flag was introduced to make the flag incrementing. I would prefer keep this new mechanism for stability.

File: csrc/custom_all_reduce_test.cu [hunk 2]
```diff
@@ -330,10 +330,17 @@ int main(int argc, char** argv) {
   //     run<half>(myRank, nRanks, comm, threads, block_limit, 4096 * 1024);
   //   }
   // }
+#ifdef USE _ROCM
```
Comment (mawong-amd) [inline csrc/custom_all_reduce_test.cu:333]: I understand that #137 fixed this typo, but this should cause vLLM compilation to fail. How was this PR merged when it doesn't even compile?

Comment (mawong-amd) [inline csrc/custom_all_reduce_test.cu:333]: I understand this is fixed in #137, but this  should easily not pass review. Let's spend more time ensuring code quality and that tests pass before merging.

**Comment by mawong-amd** [comment]
> Hey Folks..wanted to make a quick comment. This CAR will require rocm 6.2 to compile.  the scoping intrinsics are introduced in LLVM in the 6.2 release. So we may see compile errors on `__MEMORY_SCOPE_DEVICE` etc.

We should not push out releases where the default settings (ROCm 6.1) do not compile. Again, I understand this is already fixed by #137 but hotfixes should be kept to a minimum. Especially when this issue is so readily detectable.

**Pull request merged**